### PR TITLE
[fix]: fixes to temporal filters from erratum

### PR DESCRIPTION
**Pull request by and-rewsmith**
Fixing various things inspired by the erratum:
1. Fix to [base learning rule](https://github.com/fmi-basel/latent-predictive-learning/blob/main/erratum/erratum.md#implementation-of-transmitter-triggered-plasticity-issue-2)
2. Fix to [double exponential filter](https://github.com/fmi-basel/latent-predictive-learning/blob/main/erratum/erratum.md#implementation-of-double-exponential-filtering-for-synaptic-traces-issue-3)
3. Fix to [constants](https://github.com/fmi-basel/latent-predictive-learning/blob/main/erratum/erratum.md#erroneous-sigma2-trace-normalization-issue-4)

Wrapped with this change is:
1. A rename of the `TemporalFilter` to `DoubleExponentialFilter`, as there are now many temporal filters so a bad name.
2. A conversion of the euler discretization to exponential form (more accurate)

### CODE REVIEW

File: model/src/util.py [hunk 1]
```diff
@@ -96,7 +96,8 @@ def apply(self, value: torch.Tensor, dt: float = DT) -> torch.Tensor:
             self.fall = torch.zeros_like(value)
 
         # Apply the exponential decay to the rise state and add the error
-        self.rise = self.rise * math.exp(-dt / self.tau_rise) + value
+        # NOTE: This was corrected in the erratum. Value scaled by `tau_rise`.
+        self.rise = self.rise * math.exp(-dt / self.tau_rise) + self.tau_rise * value
```
Comment (and-rewsmith) [inline model/src/util.py:100]: I believe this is what the erratum implied but happy to be corrected.

File: model/src/util.py [hunk 2]
```diff
@@ -135,7 +136,8 @@ def apply(self, spike: torch.Tensor, dt: float = DT) -> torch.Tensor:
             self.mean = torch.zeros_like(spike)
 
         # Apply the exponential decay to the mean state and add the new spike value
-        self.mean += (dt / self.tau_mean) * (spike - self.mean)
+        self.mean = self.mean * math.exp(-dt / self.tau_mean) + spike
```
Comment (and-rewsmith) [inline model/src/util.py:139]: Switching to more accurate discretization.

File: model/src/util.py [hunk 3]
```diff
@@ -170,8 +172,8 @@ def apply(self, spike: torch.Tensor, spike_moving_average: torch.Tensor, dt: flo
             self.variance = torch.zeros_like(spike)
 
         # Apply the exponential decay to the variance state and add the squared deviation
-        self.variance += (dt / self.tau_var) * \
-            ((spike - spike_moving_average) ** 2 - self.variance)
+        self.variance = self.variance * math.exp(-dt / self.tau_var) + \
```
Comment (and-rewsmith) [inline model/src/util.py:175]: Switching to more accurate discretization.

File: model/src/util.py [hunk 4]
```diff
@@ -196,10 +198,10 @@ def __init__(self, tau_stdp: float = TAU_STDP) -> None:
 
     def apply(self, spike: torch.Tensor, dt: float = DT) -> torch.Tensor:
         if self.trace is None:
-            # Initialize variance based on the first spike received
+            # Initialize trace based on the first spike received
             self.trace = torch.zeros_like(spike)
 
-        self.trace += (-1) * (dt / self.tau_stdp) * self.trace + spike
+        self.trace = self.trace * math.exp(-dt / self.tau_stdp) + self.tau_stdp * spike
```
Comment (and-rewsmith) [inline model/src/util.py:204]: Switching to more accurate discretization.

File: model/src/util.py [hunk 5]
```diff
@@ -96,10 +96,12 @@ def apply(self, value: torch.Tensor, dt: float = DT) -> torch.Tensor:
...
 
         # Apply the exponential decay to the rise state and add the error
-        self.rise = self.rise * math.exp(-dt / self.tau_rise) + value
+        decay_factor_rise = math.exp(-dt / self.tau_rise)
+        self.rise = self.rise * decay_factor_rise + value
 
         # Apply the exponential decay to the fall state and add the rise state
-        self.fall = self.fall * math.exp(-dt / self.tau_fall) + self.rise
+        decay_factor_fall = math.exp(-dt / self.tau_fall)
+        self.fall = self.fall * decay_factor_fall + (1 - decay_factor_fall) * self.rise
```
Comment (and-rewsmith) [inline model/src/util.py:104]: I believe this is what the erratum implied but happy to be corrected.

File: model/src/util.py [hunk 6]
```diff
@@ -135,7 +137,9 @@ def apply(self, spike: torch.Tensor, dt: float = DT) -> torch.Tensor:
             self.mean = torch.zeros_like(spike)
 
         # Apply the exponential decay to the mean state and add the new spike value
-        self.mean += (dt / self.tau_mean) * (spike - self.mean)
+        decay_factor = math.exp(-dt / self.tau_mean)
+        self.mean = self.mean * decay_factor + (1 - decay_factor) * spike
```
Comment (and-rewsmith) [inline model/src/util.py:141]: Switching to more accurate discretization.

File: model/src/util.py [hunk 7]
```diff
@@ -170,8 +174,9 @@ def apply(self, spike: torch.Tensor, spike_moving_average: torch.Tensor, dt: flo
             self.variance = torch.zeros_like(spike)
 
         # Apply the exponential decay to the variance state and add the squared deviation
-        self.variance += (dt / self.tau_var) * \
-            ((spike - spike_moving_average) ** 2 - self.variance)
+        decay_factor = math.exp(-dt / self.tau_var)
+        self.variance = self.variance * decay_factor + \
```
Comment (and-rewsmith) [inline model/src/util.py:178]: Switching to more accurate discretization.

File: model/tests/test_util.py [hunk 8]
```diff
@@ -1,35 +1,35 @@
...
 
     # Apply a single error with a small dt should result in small change
     assert tf.apply(value=torch.Tensor([1]), dt=1).item(
-    ) == pytest.approx(1.7357588823428847)
+    ) == pytest.approx(1.097208857536316)
 
     # Apply a zero error should result in decay of the state
     assert tf.apply(value=torch.Tensor([0]), dt=1).item(
-    ) == pytest.approx(1.1417647320527227)
+    ) == pytest.approx(0.7217329740524292)
```
Comment (and-rewsmith) [inline model/tests/test_util.py:19]: Now these test values make more sense.

**Comment by and-rewsmith** [comment]
I'll want to discuss the graphs for this model at some point. The second term keeps growing while the prefilter value is never below 1. I think it may have something to do with the implementation of some of these temporal filters.

**Comment by mstormbull** [comment]
> I'll want to discuss the graphs for this model at some point. The second term keeps growing while the prefilter value is never below 1. I think it may have something to do with the implementation of some of these temporal filters.

I am to believe that you do not get a curve like this? 
image

Happy to discuss! Sorry the timezone and conference commitments are making it harder to catch up and sync.

**Comment by and-rewsmith** [comment]
@mstormbull Yes I don't get something like that, but this weekend I have changed a lot of the internal structure to the model. Let's reassess this once these latest changes go in. I am sending a mail soon which will outline what is in the next PR.
